synapse/handlers/message.py

```diff
@@ -859,10 +824,38 @@ async def handle_new_client_event(
             ratelimit
             extra_users: Any extra users to notify about event
 
+            ignore_shadow_ban: True if shadow-banned users should be allowed to
+                send this event.
+
         Return:
-            The stream_id of the persisted event.
+            If the event was deduplicated, the previous, duplicate, event. Otherwise,
+            `event`.
+
+        Raises:
+            ShadowBanError if the requester has been shadow-banned.
         """
 
+        # we don't apply shadow-banning to membership events, so that the user
+        # can come and go as they want.
```

Comment:
I'm not sure this is really true? We do have code to handle them for invites, but I suppose those are handled before this point is reached.

Reply:
I guess... I'll add a couple more words.